File: core/src/main/scala/com/sksamuel/eel/domain.scala
package com.sksamuel.eel

import scala.language.implicitConversions

object Field {
  implicit def toField(str: String): Field = Field(str)
}

case class Field(value: String)

case class Column(name: String)

object Column {
  implicit def toField(str: String): Column = Column(str)
}

object Row {
  def apply(map: Map[String, String]): Row = {
    Row(map.keys.map(Column.apply).toSeq, map.values.seq.map(Field.apply).toSeq)
  }
}

case class Row(columns: Seq[Column], fields: Seq[Field]) {
  require(columns.size == fields.size, "Columns and fields should have the same size")

  def apply(name: String): String = {
    val pos = columns.indexWhere(_.name == name)
    fields(pos).value
  }

  def size: Int = columns.size

  def addColumn(name: String, value: String): Row = {
    copy(columns = columns :+ Column(name), fields = fields :+ Field(value))
  }

  def removeColumn(name: String): Row = Row(toMap - name)

  def toMap: Map[String, String] = columns.map(_.name).zip(fields.map(_.value)).toMap
}

File: eel-core/src/main/kotlin/io/eels/domain.kt
package io.eels

import io.eels.schema.Field
import io.eels.schema.Schema

data class Row(val schema: Schema, val values: List<Any?>) {

  init {
    require(schema.size() == values.size, { "Row should have a value for each column (${schema.fields.size} columns=${schema.fields.joinToString { "," }}, ${values.size} values=${values.joinToString { "," }})" })
  }

  override fun toString(): String {
    return schema.columnNames().zip(values).map { it ->
      "${it.first} = ${if (it.second == null) "" else it.second.toString()}"
    }.joinToString ("[", ",", "]")
  }

  fun get(k: Int): Any? = values[k]

  fun get(name: String, caseInsensitive: Boolean = false): Any? {
    val index = schema.indexOf(name, caseInsensitive)
    return values[index]
  }

  companion object {
    val PoisonPill = Row(Schema(Field("a")), listOf(object : Any() {}))
  }

  fun size(): Int = values.size

  fun replace(name: String, value: Any, caseSensitive: Boolean): Row {
    val k = schema.indexOf(name, caseSensitive)
    // todo this could be optimized to avoid the copy
    val newValues = values.toMutableList()
    newValues[k] = value
    return copy(values = newValues.toList())
  }

  fun add(name: String, value: Any): Row = copy(schema = schema.addColumn(name), values = values + value)
}